File: 20.valid-parentheses.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;

/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {
    public boolean isValid(String s) {
        ArrayList<String> stack = new ArrayList<String>();
        HashMap<String, String> brackets = new HashMap<String, String>() {
            {
                put("(", ")");
                put("{", "}");
                put("[", "]");
            }
        };

        for (int i=0; i<s.length(); i++) {
            String cur = String.valueOf(s.charAt(i));
            if (brackets.containsKey(cur)) {
                stack.add(cur);
            } else {
                int size = stack.size();
                if (size == 0) {
                    return false;
                }
                String peak = stack.get(size - 1);
                stack.remove(size - 1);
                if (!cur.equals(brackets.get(peak))) {
                    return false;
                }
            }
   
        }

        return stack.size() == 0;
    }

    public static void main(String []args) {
        Solution s = new Solution();
        s.isValid("()");

    }
}
// @lc code=end

File: easy/20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#

# @lc code=start
class Solution:
    def isValid(self, s: str) -> bool:
        if len(s) & 1 == 1:
            return False
        
        brackets = {
            '(' : ')',
            '{' : '}',
            '[' : ']'
        }

        stack = []
        for c in s:
            if c in brackets.keys():
                stack.append(c)
            else:
                if not stack:
                    return False
                
                top = stack.pop()
                if brackets[top] != c:
                    return False

        return not stack
    
    def isValid2(self, s: str) -> bool:
        while len(s) > 0:
            l = len(s)
            s = s.replace('()','').replace('{}','').replace('[]','')
            if l==len(s): return False
        return True

# @lc code=end

